### PR TITLE
Created Stackdriver notification template

### LINKED ISSUE #432: Stackdriver notification template
Support CRUD of Stackdriver notifications as a CFT template.
https://cloud.google.com/monitoring/api/ref_v3/rest/v3/projects.notificationChannels

### PR DESCRIPTION
**Pull request by kopachevsky**
* Fix #432

### CODE REVIEW

File: dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    properties = context.properties
+    project_id = properties.get('project', context.env['project'])
+    notification_channels = properties["notificationChannels"]
+    policies = properties["policies"]
+
+    notification_channel_names = []
+
+    for index, channel in enumerate(notification_channels):
+        channel_type = channel['typeName']
+        notification_channel_name = '{}-{}'.format(channel_type, index)
```
Comment (nick4fake) [inline dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py]: Please use resource name in prefix.
context.env['name']

File: dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+                'type': channel_type,
+                'enabled': True,
+                'displayName': channel['displayName'],
+                'labels': channel['labels']
+            }
+        })
+
+    for index, policy in enumerate(policies):
+        resources.append({
+            'name': 'alerting-policy-{}'.format(index),
```
Comment (nick4fake) [inline dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py]: same
context.env['name']

File: dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py.schema [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+imports:
+  - path: stackdriver_notification_channels.py
+
+required:
+  - notificationChannels
+  - policies
+
+properties:
+  notificationChannels:
+    type: array
```
Comment (nick4fake) [inline dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py.schema]: Please add minItems

File: dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py.schema [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    description: |
+        A NotificationChannel is a medium through which an alert is delivered
+        when a policy violation is detected. Examples of channels include email, SMS,
+        and third-party messaging applications. Fields containing sensitive information
+        like authentication tokens or contact info are only partially populated on retrieval.
+    items:
+      type: object
```
Comment (nick4fake) [inline dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py.schema]: additionalProperties: false

File: dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py.schema [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+          description: An optional human-readable description of this notification channel. This may not exceed 1024 Unicode characters.
+          type: string
+        displayName:
+          type: string
+          description: An optional human-readable name for this notification channel.
+        typeName:
+          type: string
+          description: The type of the notification channel.
+  policies:
+    type: array
```
Comment (nick4fake) [inline dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py.schema]: same

File: dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py.schema [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    description: |
+      A description of the conditions under which some aspect of your system
+      is considered to be "unhealthy" and the ways to notify people or services about
+      this state. For an overview of alert policies, see Introduction to Alerting.
+    items:
+      type: object
```
Comment (nick4fake) [inline dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py.schema]: same

File: dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py.schema [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+      properties:
+        name:
+          description: |
+            A short name or phrase used to identify the policy in dashboards,
+            notifications, and incidents. To avoid confusion, don't use the same display
+            name for multiple policies in the same project. The name is limited to 512
+            Unicode characters.
+          type: string
+        conditions:
+          type: array
```
Comment (nick4fake) [inline dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py.schema]: same

File: dm/templates/stackdriver_notification_channels/examples/stackdriver_notification_channels.yaml [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+imports:
+  - path: templates/stackdriver_notification_channels/stackdriver_notification_channels.py
+    name: stackdriver_notification_channels.py
+
+resources:
+  - name: my-notification-channels
+    type: stackdriver_notification_channels.py
+    properties:
+      notificationChannels:
+        -
```
Comment (ocsig) [inline dm/templates/stackdriver_notification_channels/examples/stackdriver_notification_channels.yaml]: Is there an accidential new line here?

File: dm/templates/stackdriver_notification_channels/README.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+## Prerequisites
+
+- Install [gcloud](https://cloud.google.com/sdk)
+- Create a [GCP project, set up billing, enable requisite APIs](../project/README.md)
+- Create a [Stackdriver Workspace](https://cloud.google.com/monitoring/workspaces/)
+- Log in to the [Stackdriver Workspace](https://cloud.google.com/monitoring/workspaces/)
+  where the metric has to be deployed
+- Grant the [monitoring.admin](https://cloud.google.com/monitoring/access-control)
+  IAM role to the Deployment Manager service account
```
Comment (ocsig) [inline dm/templates/stackdriver_notification_channels/README.md]: Do we need to enable any of these APIs?
```
cloudmonitoring.googleapis.com                        Cloud Monitoring API
monitoring.googleapis.com                             Stackdriver Monitoring API
```

File: dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+        channel_type = channel['typeName']
+        notification_channel_name = '{}-{}-{}'.format(resource_name, channel_type, index)
+        notification_channel_names.append('$(ref.{}.name)'.format(notification_channel_name))
+        resources.append({
+            'name': notification_channel_name,
+            'type': 'gcp-types/monitoring-v3:projects.notificationChannels',
+            'properties': {
+                'name': 'projects/{}'.format(project_id),
+                'type': channel_type,
+                'enabled': True,
```
Comment (ocsig) [inline dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py]: Why is this hardcoded and not wired out to the inputs with a default `True` value? Same for Policies.

File: dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        resources.append({
+            'name': 'alerting-policy-{}-{}'.format(context.env['name'], index),
+            'type': 'gcp-types/monitoring-v3:projects.alertPolicies',
+            'properties': {
+                'displayName': policy['name'],
+                'documentation': {
+                    'content': policy['documentationContent'],
+                    'mimeType': DEFAULT_MIME_TYPE
+                },
+                'combiner': 'OR',
```
Comment (ocsig) [inline dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py]: Why is `OR` hardcoded?

File: dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py.schema (hunk 7 above)
Comment (ocsig) [reply dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py.schema]: Is this fixed?

File: dm/templates/stackdriver_notification_channels/tests/integration/stackdriver_notification_channels.bats [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+########## TESTS ##########
+
+@test "Creating deployment ${DEPLOYMENT_NAME} from ${CONFIG}" {
+    run gcloud deployment-manager deployments create "${DEPLOYMENT_NAME}" \
+        --config "${CONFIG}" \
+        --project "${CLOUD_FOUNDATION_PROJECT_ID}"
+    [[ "$status" -eq 0 ]]
+}
+
```
Comment (ocsig) [inline dm/templates/stackdriver_notification_channels/tests/integration/stackdriver_notification_channels.bats]: Please add a test to check resource values and the existance of the Notification resource.

File: dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+        channel_type = channel['typeName']
+        notification_channel_name = '{}-{}-{}'.format(resource_name, channel_type, index)
+        notification_channel_names.append('$(ref.{}.name)'.format(notification_channel_name))
+        resources.append({
+            'name': notification_channel_name,
+            'type': 'gcp-types/monitoring-v3:projects.notificationChannels',
+            'properties': {
+                'name': 'projects/{}'.format(project_id),
+                'type': channel_type,
+                'enabled': channel.get('enabled', True),
```
Comment (ocsig) [inline dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py]: `enabled` is not mentioned in Schema.
Also default values should be set in schema, not deep in code. ( Same for policy)

File: dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+        resources.append({
+            'name': 'alerting-policy-{}-{}'.format(context.env['name'], index),
+            'type': 'gcp-types/monitoring-v3:projects.alertPolicies',
+            'properties': {
+                'displayName': policy['name'],
+                'documentation': {
+                    'content': policy['documentationContent'],
+                    'mimeType': policy.get('mimeType', DEFAULT_MIME_TYPE)
+                },
+                'combiner': policy.get('combiner', 'OR'),
```
Comment (ocsig) [inline dm/templates/stackdriver_notification_channels/stackdriver_notification_channels.py]: Combiner is not defined in schema. Please move to default `OR` to schema

**Comment by ocsig** [comment]
Schema is missing version.
PR is missing Changelog entry

**Pull request merged**